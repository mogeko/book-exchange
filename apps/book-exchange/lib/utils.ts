import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tap<T>(fn: (x: T) => void) {
  return (x: T) => (fn(x), x);
}

export function range(to: number): number[];
export function range(from: number, to: number): number[];
export function range(from: number, to?: number) {
  return Array.from({ length: to ? to - from : from }, (_, i) => i + from);
}
