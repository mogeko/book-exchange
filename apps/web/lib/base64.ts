export { decode, encode } from "js-base64";
