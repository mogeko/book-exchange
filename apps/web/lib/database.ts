import "server-only";

import { PrismaClient } from "@mogeko/bookworm-db";

declare global {
  var prisma: PrismaClient;
}

// Make sure that PrismaClient is instantiated only once.
// See: https://github.com/prisma/prisma/issues/1983#issuecomment-620621213
export const prisma = (() => {
  if (process.env.NODE_ENV === "production") {
    return new PrismaClient();
  } else {
    if (!global.prisma) {
      global.prisma = new PrismaClient();
    }

    return global.prisma;
  }
})();

export * from "@mogeko/bookworm-db";
