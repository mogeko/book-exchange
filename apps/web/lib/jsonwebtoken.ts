import { jwtVerify, SignJWT, type JWTPayload } from "jose";

import { secret } from "@/config/secret";

export async function sign(payland: JWTPayload, { expiresIn, alg }: Opts = {}) {
  return await new SignJWT(payland)
    .setProtectedHeader({ alg: alg ?? "HS256" })
    .setExpirationTime(expiresIn ?? "7d")
    .sign(new TextEncoder().encode(secret.jwt));
}

export async function verify(token: string, { alg }: Pick<Opts, "alg"> = {}) {
  const key = new TextEncoder().encode(secret.jwt);
  const { payload } = await jwtVerify(token, key, {
    algorithms: [alg ?? "HS256"],
  });

  return payload;
}

type Opts = {
  expiresIn?: string | number;
  alg?: string;
};
