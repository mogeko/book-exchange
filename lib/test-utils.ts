import { render, renderHook } from "@testing-library/react";
import type { RenderOptions, RenderHookOptions } from "@testing-library/react";
import WrapProvider from "@/providers/providers";
import { SWRConfig } from "swr";
import React from "react";

function renderWithProvider(ui: React.ReactElement, options?: RenderOptions) {
  return render(ui, { wrapper: WrapProviderForTest, ...options });
}

function renderHookWithProvider<Result, Props>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props>
) {
  return renderHook(render, { wrapper: WrapProviderForTest, ...options });
}

const WrapProviderForTest: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  return React.createElement(
    WrapProvider,
    null,
    React.createElement(
      SWRConfig,
      { value: { provider: () => new Map() } },
      children
    )
  );
};

export * from "@testing-library/react";
export { renderHookWithProvider as renderHook };
export { renderWithProvider as render };
