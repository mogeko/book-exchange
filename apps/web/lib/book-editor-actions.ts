"use server";

import { revalidatePath } from "next/cache";

import { prisma } from "@/lib/database";
import type { DataTypes } from "@/components/book-editor";

export async function updateBook(data: DataTypes) {
  try {
    const { id: bid } = await prisma.book.upsert({
      where: { title: data.title },
      update: {
        title: data.title,
        subtitle: data.subtitle,
        description: data.description,
        cover: data.cover,
        authors: {
          connectOrCreate: data?.authors?.map((author) => ({
            where: { name: author.value },
            create: { name: author.value, avatar: "" },
          })),
        },
        translators: {
          connectOrCreate: data?.translators?.map((translator) => ({
            where: { name: translator.value },
            create: { name: translator.value, avatar: "" },
          })),
        },
      },
      create: {
        title: data.title,
        subtitle: data.subtitle,
        description: data.description,
        cover: data.cover,
        authors: {
          connectOrCreate: data?.authors?.map((author) => ({
            where: { name: author.value },
            create: { name: author.value, avatar: "" },
          })),
        },
        translators: {
          connectOrCreate: data?.translators?.map((translator) => ({
            where: { name: translator.value },
            create: { name: translator.value, avatar: "" },
          })),
        },
      },
    });

    return revalidatePath(`/book/${bid}`), {};
  } catch (error: any) {
    return { error: error.message };
  }
}
