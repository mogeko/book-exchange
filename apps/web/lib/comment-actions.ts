"use server";

import { revalidatePath } from "next/cache";

import { prisma, type Comment } from "@/lib/database";
import type { VoteState } from "@/components/comment-context";

export async function removeComment(cid: number) {
  try {
    const comment = await prisma.comment.delete({ where: { id: cid } });

    return revalidatePathHelper(comment);
  } catch (error: any) {
    return { error: error.message };
  }
}

export async function likeDislike(state: VoteState, uid: number, cid: number) {
  console.log(state, cid, uid);
  try {
    if (state) {
      const { comment } = await prisma.voter.upsert({
        where: { voterId_commentId: { commentId: cid, voterId: uid } },
        update: { vote: state },
        create: {
          voter: { connect: { id: uid } },
          comment: { connect: { id: cid } },
          vote: state,
        },
        include: { comment: true },
      });

      return revalidatePathHelper(comment);
    } else {
      const { comment } = await prisma.voter.delete({
        where: { voterId_commentId: { commentId: cid, voterId: uid } },
        include: { comment: true },
      });

      return revalidatePathHelper(comment);
    }
  } catch (error: any) {
    return { error: error.message };
  }
}

const revalidatePathHelper = (comment: Comment) => {
  if (comment.authorId) {
    revalidatePath(`/auther/${comment.authorId}`);
  } else if (comment.publisherId) {
    revalidatePath(`/publisher/${comment.publisherId}`);
  } else if (comment.seriesId) {
    revalidatePath(`/series/${comment.seriesId}`);
  } else if (comment.userId) {
    revalidatePath(`/user/${comment.userId}`);
  }

  return {};
};
