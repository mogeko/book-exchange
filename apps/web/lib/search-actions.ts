"use server";

import { prisma } from "@/lib/database";

export async function search(query: string) {
  return await prisma.book.findMany({
    where: {
      OR: [
        { title: { contains: query } },
        { authors: { some: { name: { contains: query } } } },
        { isbn: { contains: query } },
      ],
    },
    select: { id: true, title: true },
    orderBy: {
      owners: { _count: "desc" },
    },
    take: 10,
    skip: 0,
  });
}
