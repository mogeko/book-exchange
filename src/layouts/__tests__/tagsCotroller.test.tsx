import "@testing-library/jest-dom";
import { render, screen, waitFor } from "@/lib/test-utils";
import TagsCotroller from "@/layouts/tagsCotroller";

describe("TagsCotroller", () => {
  it("should render correctly", async () => {
    const { container } = render(<TagsCotroller />);

    expect(screen.getByRole("heading", { name: /tags/i })).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.getAllByRole("heading").length).toBeGreaterThanOrEqual(4);
      expect(
        screen.getAllByRole("link", { name: /tag/i }).length
      ).toBeGreaterThanOrEqual(3);
    });

    expect(container).toMatchSnapshot();
  });
});
