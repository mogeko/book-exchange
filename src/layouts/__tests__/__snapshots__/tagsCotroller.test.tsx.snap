// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsCotroller should render correctly 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <h1
      class="text-2xl w-full py-2"
    >
      Tags
    </h1>
    <div
      class="flex flex-col items-start my-2"
    >
      <h2
        class="text-lg capitalize w-full my-2"
      >
        Fresh
      </h2>
      <div
        class="flex flex-wrap gap-2"
      >
        <a
          aria-label="tag"
          class="btn btn-xs normal-case"
          href="/tags/portals"
        >
          portals
        </a>
        <a
          aria-label="tag"
          class="btn btn-xs normal-case"
          href="/tags/portals"
        >
          portals
        </a>
        <a
          aria-label="tag"
          class="btn btn-xs normal-case"
          href="/tags/portals"
        >
          portals
        </a>
      </div>
    </div>
  </div>
</div>
`;
